### DISCUSSION TITLE
PETSc Preconditioners

### DISCUSSION THREAD

**Discussion by TLWise** [idaholab/moose#27299] [Q&A General] [answered]
PETSc appears to have many options for preconditioners. Are all of the PETSc preconditioners implemented for use in MOOSE?

**Comment by lindsayad** [comment]
We do very little hard-coding of solve related options in MOOSE related to PETSc such that users should have almost complete accessibility to PETSc preconditioning options through PETSc command line options, e.g. `-pc_type hypre -pc_hypre_type boomeramg`, `-pc_type gamg`, `-pc_type lu`, `-pc_type asm`, `-pc_type bjacobi`, `-pc_type jacobi`, so on and so forth

**Reply by TLWise** [reply]
@lindsayad PETSc user manual has the following statement, "A large suite of routines is available for using geometric multigrid as a preconditioner" 

It appears, though, that the implementation of this methodology is more involved than using predefined syntax. They follow on with the statement, "A multigrid preconditioner is created with the four commands

KSPCreate(MPI_Comm comm,KSP *ksp);
KSPGetPC(KSP ksp,PC *pc);
PCSetType(PC pc,PCMG);
PCMGSetLevels(pc,PetscInt levels,MPI_Comm *comms);

A large number of parameters affect the multigrid behavior." 

I was not sure if the above steps have been coded into moose, allowing the moose user access to the multigrid methods referred to.

**Reply by lindsayad** [reply, marked as answer]
I should have known you'd go straight to the hard one 😄 I should perhaps say we have complete accessibility to algebraic preconditioners, including algebraic multigrid, but we do not have the hooks currently for geometric multigrid. Our parent library libMesh has demonstrated geometric multigrid (see https://github.com/libMesh/libmesh/issues/1567 and https://github.com/libMesh/libmesh/pull/2047) but we have not done so in MOOSE yet. However, doing so is a desire of @roystgnr and mine. I just wrote it into our NEAMS 5-year plan ... so it's probably not going to happen immediately
